### ISSUE TITLE
Virtual Microphone not showing up in Devices

### ISSUE THREAD

**Issue by Abhinickz** [lawl/NoiseTorch#159]
I am running noisetorch `0.11.3` with pipewire `0.3.31`, found no error on the log but the virtual microphone is not showing up in devices.

```
Operating System: Ubuntu 21.04
KDE Plasma Version: 5.21.4
```

![audio_devices](https://user-images.githubusercontent.com/10339494/123935211-a2e09c80-d9b1-11eb-9d29-5fac6ed53806.png)
![image](https://user-images.githubusercontent.com/10339494/123935345-c4418880-d9b1-11eb-9f59-d03fd4bcc2ac.png)

```
$ noisetorch -log
2021/06/30 14:36:44 Application starting. Version: 0.11.3 (official)
2021/06/30 14:36:44 CAP_SYS_RESOURCE: true
2021/06/30 14:36:44 Checking if config needs to be initialized
2021/06/30 14:36:44 Couldn't resolve $XDG_CONFIG_HOME falling back to '/home/user/.config'
2021/06/30 14:36:44 Wrote temp librnnoise to: /tmp/librnnoise-130660008.so
2021/06/30 14:36:44 Couldn't resolve $XDG_CONFIG_HOME falling back to '/home/user/.config'
2021/06/30 14:36:44 Audioserver package name: PulseAudio (on PipeWire 0.3.31)
2021/06/30 14:36:44 Audioserver package version: 14.0.0
2021/06/30 14:36:44 Detected PipeWire
2021/06/30 14:36:44 Getting caps for: /home/user/.local/bin/noisetorch
2021/06/30 14:36:44 Audioserver package name: PulseAudio (on PipeWire 0.3.31)
2021/06/30 14:36:44 Audioserver package version: 14.0.0
2021/06/30 14:36:44 Detected PipeWire
2021/06/30 14:36:44 Connected to audio server. Server name 'PipeWire'
2021/06/30 14:36:44 Output alsa_output.pci-0000_00_1f.3.analog-stereo, {Index:43 Name:alsa_output.pci-0000_00_1f.3.analog-stereo Description:Built-in Audio Analog Stereo SampleSpec:{Format:7 Channels:2 Rate:48000} ChannelMap:[1 2] ModuleIndex:4294967295 Cvolume:[65536 65536] Muted:false MonitorSourceIndex:65579 MonitorSourceName:alsa_output.pci-0000_00_1f.3.analog-stereo.monitor Latency:0 Driver:PipeWire Flags:183 PropList:map[alsa.card:0 alsa.card_name:HDA Intel PCH alsa.class:generic alsa.device:0 alsa.driver_name:snd_hda_intel alsa.id:ALC298 Analog alsa.long_card_name:HDA Intel PCH at 0xf1240000 irq 140 alsa.name:ALC298 Analog alsa.resolution_bits:16 alsa.subclass:generic-mix alsa.subdevice:0 alsa.subdevice_name:subdevice #0 api.alsa.path:front:0 api.alsa.pcm.card:0 api.alsa.pcm.stream:playback audio.adapt.follower: audio.channels:2 audio.position:FL,FR card.profile.device:8 client.id:32 device.api:alsa device.class:sound device.description:Built-in Audio Analog Stereo device.icon_name:audio-card-analog device.id:40 device.profile.description:Analog Stereo device.profile.name:analog-stereo device.routes:2 factory.id:18 factory.mode:merge factory.name:api.alsa.pcm.sink library.name:audioconvert/libspa-audioconvert media.class:Audio/Sink node.driver:true node.max-latency:8192/48000 node.name:alsa_output.pci-0000_00_1f.3.analog-stereo node.nick:HDA Intel PCH node.pause-on-idle:false object.id:43 object.path:alsa:pcm:0:front:0:playback priority.driver:1009 priority.session:1009] RequestedLatency:0 BaseVolume:65536 SinkState:2 NVolumeSteps:0 CardIndex:40 Ports:[{Name:analog-output-speaker Description:Speakers Pririty:10000 Available:0} {Name:analog-output-headphones Description:Headphones Pririty:9900 Available:1}] ActivePortName:analog-output-speaker Formats:[{Encoding:1 PropList:map[]}]}
2021/06/30 14:36:44 Output bluez_output.38_18_4C_49_48_59.a2dp-sink, {Index:84 Name:bluez_output.38_18_4C_49_48_59.a2dp-sink Description:WH-XB900N SampleSpec:{Format:5 Channels:2 Rate:48000} ChannelMap:[1 2] ModuleIndex:4294967295 Cvolume:[43346 43346] Muted:false MonitorSourceIndex:65620 MonitorSourceName:bluez_output.38_18_4C_49_48_59.a2dp-sink.monitor Latency:0 Driver:PipeWire Flags:167 PropList:map[api.bluez5.address:38:18:4C:49:48:59 api.bluez5.codec:ldac api.bluez5.profile:a2dp-sink api.bluez5.transport: audio.adapt.follower: card.profile.device:1 client.id:31 device.api:bluez5 device.bus:bluetooth device.description:WH-XB900N device.form_factor:headset device.icon_name:audio-headset-bluetooth device.id:75 device.routes:1 factory.id:8 factory.mode:merge factory.name:api.bluez5.a2dp.sink library.name:audioconvert/libspa-audioconvert media.class:Audio/Sink node.driver:true node.latency:512/48000 node.name:bluez_output.38_18_4C_49_48_59.a2dp-sink node.pause-on-idle:false object.id:84 priority.driver:1010 priority.session:1010] RequestedLatency:0 BaseVolume:65536 SinkState:2 NVolumeSteps:512 CardIndex:75 Ports:[{Name:headset-output Description:Headset Pririty:0 Available:2}] ActivePortName:headset-output Formats:[{Encoding:1 PropList:map[]}]}
2021/06/30 14:36:47 Couldn't resolve $XDG_CONFIG_HOME falling back to '/home/user/.config'
2021/06/30 14:36:47 Couldn't resolve $XDG_CONFIG_HOME falling back to '/home/user/.config'
2021/06/30 14:36:48 Couldn't resolve $XDG_CONFIG_HOME falling back to '/home/user/.config'
2021/06/30 14:36:48 Couldn't resolve $XDG_CONFIG_HOME falling back to '/home/user/.config'
2021/06/30 14:36:48 Couldn't resolve $XDG_CONFIG_HOME falling back to '/home/user/.config'
2021/06/30 14:36:48 Couldn't resolve $XDG_CONFIG_HOME falling back to '/home/user/.config'
2021/06/30 14:36:48 Couldn't resolve $XDG_CONFIG_HOME falling back to '/home/user/.config'
2021/06/30 14:36:48 Couldn't resolve $XDG_CONFIG_HOME falling back to '/home/user/.config'
2021/06/30 14:36:50 Unloading modules for pipewire
2021/06/30 14:36:50 Searching for module-ladspa-source
2021/06/30 14:36:50 Found module-ladspa-source at id [262144], sending unload command
2021/06/30 14:36:50 Searching for module-ladspa-sink
2021/06/30 14:36:50 Loading supressor for pipewire
2021/06/30 14:36:50 Loaded ladspa source as idx: 262144
2021/06/30 14:36:50 Couldn't resolve $XDG_CONFIG_HOME falling back to '/home/user/.config'
```

**Comment by lawl** [comment]
> not showing up in devices

I don't know what "devices" is. i don't recognize the application in the screenshot, is that a KDE thing? Does it show up in pavucontrol?

**Comment by Abhinickz** [comment]
> I don't know what "devices" is. i don't recognize the application in the screenshot, is that a KDE thing? Does it show up in pavucontrol?

Yes, that is the KDE Audio Setting and by device, I meant as an Input device (Microphone).

Like in Gnome this is shown:



![](https://1.bp.blogspot.com/-_qSrjp9sk08/YDOp05dz6rI/AAAAAAAAFRQ/Hr0EBtW55n4BjAMD06lnGBVkgptWaM8OgCLcBGAsYHQ/s1050/noisetorch-system-settings.png)

**Comment by lawl** [comment]
> Does it show up in pavucontrol?

**Comment by lawl** [comment]
Can you load it again, and then run

     pactl list sources

and attach the output here?

**Comment by Abhinickz** [comment]
After loading it again same behaviour:
```
$ pactl list sources
Source #65690
        State: SUSPENDED
        Name: alsa_output.pci-0000_00_1f.3.analog-stereo.monitor
        Description: Monitor of Built-in Audio Analog Stereo
        Driver: PipeWire
        Sample Specification: s32le 2ch 48000Hz
        Channel Map: front-left,front-right
        Owner Module: 4294967295
        Mute: no
        Volume: front-left: 65536 / 100% / 0.00 dB,   front-right: 65536 / 100% / 0.00 dB
                balance 0.00
        Base Volume: 65536 / 100% / 0.00 dB
        Monitor of Sink: alsa_output.pci-0000_00_1f.3.analog-stereo
        Latency: 0 usec, configured 0 usec
        Flags: HARDWARE DECIBEL_VOLUME LATENCY 
        Properties:
                object.path = "alsa:pcm:0:front:0:playback"
                api.alsa.path = "front:0"
                api.alsa.pcm.card = "0"
                api.alsa.pcm.stream = "playback"
                audio.channels = "2"
                audio.position = "FL,FR"
                device.routes = "2"
                alsa.resolution_bits = "16"
                device.api = "alsa"
                device.class = "sound"
                alsa.class = "generic"
                alsa.subclass = "generic-mix"
                alsa.name = "ALC298 Analog"
                alsa.id = "ALC298 Analog"
                alsa.subdevice = "0"
                alsa.subdevice_name = "subdevice #0"
                alsa.device = "0"
                alsa.card = "0"
                alsa.card_name = "HDA Intel PCH"
                alsa.long_card_name = "HDA Intel PCH at 0xf1240000 irq 140"
                alsa.driver_name = "snd_hda_intel"
                device.profile.name = "analog-stereo"
                device.profile.description = "Analog Stereo"
                card.profile.device = "8"
                device.id = "40"
                factory.name = "api.alsa.pcm.sink"
                priority.driver = "1009"
                priority.session = "1009"
                media.class = "Audio/Sink"
                node.nick = "HDA Intel PCH"
                node.name = "alsa_output.pci-0000_00_1f.3.analog-stereo"
                device.description = "Built-in Audio Analog Stereo"
                device.icon_name = "audio-card-analog"
                node.pause-on-idle = "false"
                factory.id = "18"
                client.id = "32"
                node.driver = "true"
                factory.mode = "merge"
                audio.adapt.follower = ""
                library.name = "audioconvert/libspa-audioconvert"
                object.id = "154"
                node.max-latency = "8192/48000"
        Formats:
                pcm

Source #159
        State: SUSPENDED
        Name: alsa_input.pci-0000_00_1f.3.analog-stereo
        Description: Built-in Audio Analog Stereo
        Driver: PipeWire
        Sample Specification: s32le 2ch 48000Hz
        Channel Map: front-left,front-right
        Owner Module: 4294967295
        Mute: yes
        Volume: front-left: 13878 /  21% / -40.45 dB,   front-right: 13878 /  21% / -40.45 dB
                balance 0.00
        Base Volume: 6554 /  10% / -60.00 dB
        Monitor of Sink: n/a
        Latency: 0 usec, configured 0 usec
        Flags: HARDWARE HW_MUTE_CTRL HW_VOLUME_CTRL DECIBEL_VOLUME LATENCY 
        Properties:
                object.path = "alsa:pcm:0:front:0:capture"
                api.alsa.path = "front:0"
                api.alsa.pcm.card = "0"
                api.alsa.pcm.stream = "capture"
                audio.channels = "2"
                audio.position = "FL,FR"
                device.routes = "3"
                alsa.resolution_bits = "16"
                device.api = "alsa"
                device.class = "sound"
                alsa.class = "generic"
                alsa.subclass = "generic-mix"
                alsa.name = "ALC298 Analog"
                alsa.id = "ALC298 Analog"
                alsa.subdevice = "0"
                alsa.subdevice_name = "subdevice #0"
                alsa.device = "0"
                alsa.card = "0"
                alsa.card_name = "HDA Intel PCH"
                alsa.long_card_name = "HDA Intel PCH at 0xf1240000 irq 140"
                alsa.driver_name = "snd_hda_intel"
                device.profile.name = "analog-stereo"
                device.profile.description = "Analog Stereo"
                card.profile.device = "0"
                device.id = "40"
                factory.name = "api.alsa.pcm.source"
                priority.driver = "2009"
                priority.session = "2009"
                media.class = "Audio/Source"
                node.nick = "HDA Intel PCH"
                node.name = "alsa_input.pci-0000_00_1f.3.analog-stereo"
                device.description = "Built-in Audio Analog Stereo"
                device.icon_name = "audio-input-microphone-analog"
                node.pause-on-idle = "false"
                factory.id = "18"
                client.id = "32"
                node.driver = "true"
                factory.mode = "split"
                audio.adapt.follower = ""
                library.name = "audioconvert/libspa-audioconvert"
                object.id = "159"
                node.max-latency = "8192/48000"
        Ports:
                analog-input-internal-mic: Internal Microphone (type: Mic, priority: 8900, availability group: Legacy 1, availability unknown)
                analog-input-dock-mic: Dock Microphone (type: Mic, priority: 7800, availability group: Legacy 2, not available)
                analog-input-mic: Microphone (type: Mic, priority: 8700, availability group: Legacy 3, not available)
        Active Port: analog-input-internal-mic
        Formats:
                pcm

Source #65680
        State: SUSPENDED
        Name: bluez_output.38_18_4C_49_48_59.a2dp-sink.monitor
        Description: Monitor of WH-XB900N
        Driver: PipeWire
        Sample Specification: float32le 2ch 48000Hz
        Channel Map: front-left,front-right
        Owner Module: 4294967295
        Mute: no
        Volume: front-left: 65536 / 100% / 0.00 dB,   front-right: 65536 / 100% / 0.00 dB
                balance 0.00
        Base Volume: 65536 / 100% / 0.00 dB
        Monitor of Sink: bluez_output.38_18_4C_49_48_59.a2dp-sink
        Latency: 0 usec, configured 0 usec
        Flags: HARDWARE DECIBEL_VOLUME LATENCY 
        Properties:
                api.bluez5.transport = ""
                api.bluez5.profile = "a2dp-sink"
                api.bluez5.codec = "ldac"
                api.bluez5.address = "38:18:4C:49:48:59"
                device.routes = "1"
                card.profile.device = "1"
                device.form_factor = "headset"
                device.bus = "bluetooth"
                device.id = "101"
                device.description = "WH-XB900N"
                node.name = "bluez_output.38_18_4C_49_48_59.a2dp-sink"
                factory.name = "api.bluez5.a2dp.sink"
                priority.driver = "1010"
                priority.session = "1010"
                device.icon_name = "audio-headset-bluetooth"
                node.pause-on-idle = "false"
                factory.id = "8"
                device.api = "bluez5"
                media.class = "Audio/Sink"
                node.driver = "true"
                node.latency = "512/48000"
                factory.mode = "merge"
                audio.adapt.follower = ""
                library.name = "audioconvert/libspa-audioconvert"
                object.id = "144"
                client.id = "31"
        Formats:
                pcm
```

```
$ noisetorch -log
2021/07/06 22:45:36 Application starting. Version: 0.11.3 (official)
2021/07/06 22:45:36 CAP_SYS_RESOURCE: true
2021/07/06 22:45:36 Checking if config needs to be initialized
2021/07/06 22:45:36 Couldn't resolve $XDG_CONFIG_HOME falling back to '/home/user/.config'
2021/07/06 22:45:36 Wrote temp librnnoise to: /tmp/librnnoise-729493364.so
2021/07/06 22:45:36 Couldn't resolve $XDG_CONFIG_HOME falling back to '/home/user/.config'
2021/07/06 22:45:36 Audioserver package name: PulseAudio (on PipeWire 0.3.31)
2021/07/06 22:45:36 Audioserver package version: 14.0.0
2021/07/06 22:45:36 Detected PipeWire
2021/07/06 22:45:36 Getting caps for: /home/user/.local/bin/noisetorch
2021/07/06 22:45:36 Audioserver package name: PulseAudio (on PipeWire 0.3.31)
2021/07/06 22:45:36 Audioserver package version: 14.0.0
2021/07/06 22:45:36 Detected PipeWire
2021/07/06 22:45:36 Connected to audio server. Server name 'PipeWire'
2021/07/06 22:45:36 Output alsa_output.pci-0000_00_1f.3.analog-stereo, {Index:154 Name:alsa_output.pci-0000_00_1f.3.analog-stereo Description:Built-in Audio Analog Stereo SampleSpec:{Format:7 Channels:2 Rate:48000} ChannelMap:[1 2] ModuleIndex:4294967295 Cvolume:[65536 65536] Muted:false MonitorSourceIndex:65690 MonitorSourceName:alsa_output.pci-0000_00_1f.3.analog-stereo.monitor Latency:0 Driver:PipeWire Flags:183 PropList:map[alsa.card:0 alsa.card_name:HDA Intel PCH alsa.class:generic alsa.device:0 alsa.driver_name:snd_hda_intel alsa.id:ALC298 Analog alsa.long_card_name:HDA Intel PCH at 0xf1240000 irq 140 alsa.name:ALC298 Analog alsa.resolution_bits:16 alsa.subclass:generic-mix alsa.subdevice:0 alsa.subdevice_name:subdevice #0 api.alsa.path:front:0 api.alsa.pcm.card:0 api.alsa.pcm.stream:playback audio.adapt.follower: audio.channels:2 audio.position:FL,FR card.profile.device:8 client.id:32 device.api:alsa device.class:sound device.description:Built-in Audio Analog Stereo device.icon_name:audio-card-analog device.id:40 device.profile.description:Analog Stereo device.profile.name:analog-stereo device.routes:2 factory.id:18 factory.mode:merge factory.name:api.alsa.pcm.sink library.name:audioconvert/libspa-audioconvert media.class:Audio/Sink node.driver:true node.max-latency:8192/48000 node.name:alsa_output.pci-0000_00_1f.3.analog-stereo node.nick:HDA Intel PCH node.pause-on-idle:false object.id:154 object.path:alsa:pcm:0:front:0:playback priority.driver:1009 priority.session:1009] RequestedLatency:0 BaseVolume:65536 SinkState:2 NVolumeSteps:0 CardIndex:40 Ports:[{Name:analog-output-speaker Description:Speakers Pririty:10000 Available:0} {Name:analog-output-headphones Description:Headphones Pririty:9900 Available:1}] ActivePortName:analog-output-speaker Formats:[{Encoding:1 PropList:map[]}]}
2021/07/06 22:45:36 Output bluez_output.38_18_4C_49_48_59.a2dp-sink, {Index:144 Name:bluez_output.38_18_4C_49_48_59.a2dp-sink Description:WH-XB900N SampleSpec:{Format:5 Channels:2 Rate:48000} ChannelMap:[1 2] ModuleIndex:4294967295 Cvolume:[39218 39218] Muted:false MonitorSourceIndex:65680 MonitorSourceName:bluez_output.38_18_4C_49_48_59.a2dp-sink.monitor Latency:0 Driver:PipeWire Flags:167 PropList:map[api.bluez5.address:38:18:4C:49:48:59 api.bluez5.codec:ldac api.bluez5.profile:a2dp-sink api.bluez5.transport: audio.adapt.follower: card.profile.device:1 client.id:31 device.api:bluez5 device.bus:bluetooth device.description:WH-XB900N device.form_factor:headset device.icon_name:audio-headset-bluetooth device.id:101 device.routes:1 factory.id:8 factory.mode:merge factory.name:api.bluez5.a2dp.sink library.name:audioconvert/libspa-audioconvert media.class:Audio/Sink node.driver:true node.latency:512/48000 node.name:bluez_output.38_18_4C_49_48_59.a2dp-sink node.pause-on-idle:false object.id:144 priority.driver:1010 priority.session:1010] RequestedLatency:0 BaseVolume:65536 SinkState:2 NVolumeSteps:512 CardIndex:101 Ports:[{Name:headset-output Description:Headset Pririty:0 Available:2}] ActivePortName:headset-output Formats:[{Encoding:1 PropList:map[]}]}
2021/07/06 22:45:43 Unloading modules for pipewire
2021/07/06 22:45:43 Searching for module-ladspa-source
2021/07/06 22:45:43 Found module-ladspa-source at id [262144], sending unload command
2021/07/06 22:45:43 Searching for module-ladspa-sink
2021/07/06 22:45:43 Loading supressor for pipewire
2021/07/06 22:45:43 Loaded ladspa source as idx: 262144
2021/07/06 22:45:43 Couldn't resolve $XDG_CONFIG_HOME falling back to '/home/user/.config'
```

**Comment by lawl** [comment]
Cannot reproduce. If anyone has information on how to reliably reproduce this, let me know.

**Comment by optlink** [comment]
I have exactly the same behavior also on Pipewire 0.3.31 and NoiseTorch 0.11.3. This is on Gentoo.

The NoiseTorch device is created and shown in pw-dump and njconnect along with the correct links.
But not in pactl or pulsemixer.

[pactl.txt](https://github.com/lawl/NoiseTorch/files/6794259/pactl.txt)
[pw-dump.txt](https://github.com/lawl/NoiseTorch/files/6794260/pw-dump.txt)

**Comment by lawl** [comment]
Interesting, thanks. If the nodes show up in pipewire, but the device doesn't as a pulseaudio source, then i believe that's a pipewire bug, and  an issue should be filed there.

**Comment by optlink** [comment]
Pipewire issue [here](https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/1413)

**Comment by lawl** [comment]
This seems to have been fixed upstream with 0.3.34, see https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/NEWS
